rest-api and documentation: Update server_metadata.py to work with 2015.5.1 static_loader, and update documentation and vagrant files to work around an issue with m2crypto on later releases of Centos/Rhel/Fedora

joehandzik opened the pull request:
…k around m2crypto build problems, and update the salt version dependency in the requirements.txt files.

Signed-Off-By: Joe Handzik <joseph.t.handzik@hp.com>

joehandzik commented:
Hey guys,

Ran into issues setting up the Calamari development environment (one was a build issue with m2crypto, the other was a bizarre trace when trying to run salt-call to set up the postgresql database. The m2crypto fix is based on a (what turned out to be) good guess based on what the fedora scripts described here were doing: https://community.webfaction.com/questions/13828/installing-m2crypto-swig-error

The other issue didn't make a lick of sense, but updating to a more recent version of salt in requirements.txt file.

joehandzik commented:
Ah, looks like some of this may fold in with your pull request here: https://github.com/ceph/calamari/pull/273

My current environment is up and stumbling after the changes here, but it may not be perfect (the salt-call failed for a couple of items, but I suspected that was because I had started running the states by hand last night...). I'll keep poking to see if I can figure out what else may be wrong.

joehandzik commented:
This configuration is actually close for me, after hacking at the postgres config file the only issue I've got is that carbon-cache won't stay running. Still trying to figure that out.

GregMeno commented:
@joehandzik When you say this do you mean wip-f21-devmode or some combination of this PR and that? I'm guessing that you are using some older versions of RHEL/centos considering that you are modifying the python 2.6 path. Would you please share what you are targeting?

joehandzik commented:
My *this* was referring to the code I have in this PR. Haven't tried wip-f21-devmode. 

I'm using CentOS7, I just changed the 2.6 and 2.7 requirements.txt files together (maybe a bad move, not sure). I'll back out the salt version change for now, but I think the environment variable I added to the front end of the requirements install will save some poor soul like me a lot of searching at some point in time. If I have further updates that I feel confident in, I'll add them in a separate PR...

We can talk more about exactly what I'm doing on Friday, I figure I'll still be off a bit at that point.

joehandzik commented:
Good news: My workaround works.
Bad news: I don't like it. 

I tried writing the workaround more like:

*try to import _create_loader, if _create_loader isn't present import static_loader (but still let them maintain their respective names, instead of renaming static_loader as _create_loader
*try to call _create_loader, if that fails call static_loader

That code also works, but 'make lint' complains that I'm not using a method that I've imported (which is not true, but for whatever reason it sees that I'm importing static_loader but doesn't see that the _create_loader call will fail and result in a call to static_loader). Thoughts? I'll leave this as-is for now, but I don't think that's a particularly clean or readable way to work around the compatibility nightmare that salt appears to be.

joehandzik commented:
Looking into other ways to work around the m2crypto issue too, but the cleanest way so far is the documentation change that I implemented. I don't remember what we decided...do you want to hold up this PR for a different change on that issue? If so, would you prefer a third requirements.txt file? I'll keep looking, maybe there's some elegant way to embed an EV into a requirements.txt file temporarily that I'm missing...

GregMeno commented:
@joehandzik I don't think that you need embed the env var in the requirements file.

What I would try is using
http://salt.readthedocs.org/en/latest/ref/states/all/salt.states.environ.html
to add to
https://github.com/ceph/calamari/blob/master/vagrant/devmode/salt/roots/virtualenv.sls

I'll review the part server_metadata later today.

joehandzik commented:
@GregMeno Yeah, the salt setting would work for vagrant setup of devmode, but I ran into my issue with manual setup (hence changing the documentation). I can make the edit in the salt script for vagrant, but I'd like documentation for manual too (unless you really, really don't want it there). Would you prefer that I pull out the environment variable from that step, and instead add some sort of "FYI, if you run into m2crypto problems with swig add this to step 1:" kind of thing?

GregMeno commented:
@joehandzik I was thinking that both is the right answer.

joehandzik commented:
Ok, I misunderstood what my follow-on goal was then. I'll get a change into the virtualenv.sls file. I'll submit it as a separate commit unless you want it squashed with the documentation change.

GregMeno commented:
@joehandzik grains code looks fine.

joehandzik commented:
Ok, good news. Let me know if you need anything else on this one to get it pushed upstream.

dmick commented:
Can you just refresh me here: what platform requires the SWIG env var workaround?

joehandzik commented:
I'm running on Centos7, and it ended up being a known issue with using swig and m2crypto on later Rhel/Centos/Fedora releases (hence the existing fedora_setup.sh script mentioned here: https://community.webfaction.com/questions/13828/installing-m2crypto-swig-error)

GregMeno commented:
@joehandzik I'm still expecting to merge this once my testing is complete this morning.

It would be nice to understand what errors we are treating at warnings way down there.
http://swig.org/Doc1.3/Preprocessor.html#Preprocessor_warning_error
https://github.com/eventbrite/m2crypto/commit/81493da914b836fb6b5cb7f7f45299eaa9ab8d66

leaving these as reference for the next time I wonder.

GregMeno commented on vagrant/devmode/salt/roots/virtualenv.sls:
```diff
@@ -29,6 +29,7 @@ pip_pkgs:
     - activate: true
     - requirements: /home/vagrant/calamari/requirements/2.7/requirements.txt
     - download_cache: /vagrant/pip_cache
+    - env_vars: SWIG_FEATURES=-cpperraswarn
```
one thing that seems wrong about this approach is we would probably like to know if other SWIG dependent packages have errors that we transmute to warnings. i.e. this probably shouldn't apply as a blanket to all elements in requirements.txt

dmick replied:
Yep, that was one of my first comments.  It's a big hammer.  Oh well.  The right thing to do is unbundle all this crap and make it the distros' problem anyway.

joehandzik replied:
I mean, in the short term I could refactor m2crypto out of the requirements.txt file, and add a second call to pip in the salt file. Whatever you guys prefer.
